[release-7.7] Visual Studio text model tweaks

monojenkins opened the pull request:
This PR adds two tweaks to the current integration with the Visual Studio text model:

- It integrates with the `GetContent<T>` machinery so that an underlying `ITextDocument` instance can be queried from a `TextEditor` instance
- If fixes `ITextDocument` lifecycle when the parent document is disposed so that the corresponding event in `ITextDocumentFactoryService` is fired properly

Ultimately those two changes help us share our integration code between Visual Studio Windows and VSMac.

Backport of #6374.

/cc @garuma

garuma committed:
[SourceEditor] Allow querying TextEditor instances for their underlying VisualStudio's ITextDocument via GetContent

garuma committed:
[TextEditor] Properly dispose underlying VisualStudio's ITextDocument instance

This ensures that ITextDocumentFactoryService.TextDocumentDisposed event is raised properly

Therzok committed:
Fix unit tests and NRE on closing a document

garuma commented:
Context: https://github.com/KirillOsenkov/vs-editor-api/commit/9b4947b378c3ffc21c59b0197b5ee3cab3fd08d1

garuma committed:
[build] Bump vs-editor-api to 15.2.3 to bring in fix for ITextDocument saving

Related change: https://github.com/KirillOsenkov/vs-editor-api/commit/9b4947b378c3ffc21c59b0197b5ee3cab3fd08d1
